tool/tsh/db.go

```diff
@@ -675,24 +673,19 @@ func prepareLocalProxyOptions(arg *localProxyConfig) (*localProxyOpts, error) {
 		certs = nil
 	}
 
+	alpnProtocol, err := common.ToALPNProtocol(arg.routeToDatabase.Protocol)
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+
 	opts := &localProxyOpts{
-		proxyAddr:               arg.teleportClient.WebProxyAddr,
-		listener:                arg.listener,
-		protocols:               []common.Protocol{common.Protocol(arg.routeToDatabase.Protocol)},
-		insecure:                arg.cliConf.InsecureSkipVerify,
-		certs:                   certs,
-		alpnConnUpgradeRequired: alpnproxy.IsALPNConnUpgradeRequired(arg.teleportClient.WebProxyAddr, arg.cliConf.InsecureSkipVerify),
-	}
-
-	// If ALPN connection upgrade is required, explicitly use the profile CAs
-	// since the tunneled TLS routing connection serves the Host cert.
-	if opts.alpnConnUpgradeRequired {
-		profileCAs, err := utils.NewCertPoolFromPath(arg.profile.CACertPathForCluster(arg.rootClusterName))
-		if err != nil {
-			return nil, trace.Wrap(err)
-		}
-		opts.rootCAs = profileCAs
+		proxyAddr: arg.teleportClient.WebProxyAddr,
+		listener:  arg.listener,
+		protocols: []common.Protocol{alpnProtocol},
+		insecure:  arg.cliConf.InsecureSkipVerify,
+		certs:     certs,
 	}
+	opts.configOpts = append(opts.configOpts, alpnproxy.WithALPNConnUpgradeTest(arg.teleportClient.RootClusterCACertPool))
```

Comment:
nit: we still have `rootClusterName` available in callers of `prepareLocalProxyOptions`. Using the teleport client to get the root cluster CA Cert pool involves loading the core key, which when `require_session_mfa: hardware_key_touch` is set, means the user has to tap their yubikey. I think that the private key is available for ~15 seconds without requiring another tap, but this could be a UX regression (maybe). I guess I'll have to try it with a PIV build to see what happens

Reply:
I am not very familiar with the hardware key flow. Could you share the code where the tap is required in this case for the private key?

If it does require an extra tap, I feel we should resolve it separately regardless of this change since core key is loaded in some other places as well.

Reply:
Sorry I forgot to get around to testing this earlier. Just tried and it doesn't do an additional prompt, and in fact doesn't prompt at all when just loading the key to get the root cluster name.

The code for this is in `api/utils/keys/yubikey.go` which is only built when `piv` build tag is set. We only prompt when generating a new key or signing with an existing key, and we use the caching prompt policy so the yubikey doesn't prompt again for ~15 seconds.

I'll mark this resolved.